src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -76,15 +79,23 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
             _logger.Started();
         }
 
-        private async Task HandleBackgroundException(BackgroundService backgroundService)
+        private async Task TryStartBackgroundServiceAsync(
+            BackgroundService backgroundService, CancellationToken cancellationToken)
         {
             try
             {
+                // Fire IHostedService.Start
+                await backgroundService.StartAsync(cancellationToken).ConfigureAwait(false);
                 await backgroundService.ExecuteTask.ConfigureAwait(false);
             }
             catch (Exception ex)
             {
                 _logger.BackgroundServiceFaulted(ex);
+                if (_options.BackgroundServiceExceptionBehavior == BackgroundServiceExceptionBehavior.StopHost)
+                {
+                    _logger.BackgroundServiceStoppingHost(ex);
+                    _applicationLifetime.StopApplication();
```

Comment:
Does `_applicationLifetime.StopApplication();` work if we are in the middle of starting up other services? 

For example, if the first BackgroundService gets started asynchronously, and the Host goes off and is in the middle of starting a second service. Meanwhile the first BackgroundService throws an exception and we call `StopApplication()` while the Host is still starting, what is going to happen?

Reply:
This is part of the concern I raised about switching from `StopAsync` here to `_applicationLifetime.StopApplication()`. @davidfowl had some thoughts on this, but I'm not sure.

Reply:
I don't think it needs to be immediate. StopApplication needs to work in general, so that is broken, we can fix it separately.

Reply:
Do we know what the behavior is here? Can we add a test for it?

Reply:
Added a new test in 4737d98